vita3k/host/src/pup.cpp

```diff
@@ -0,0 +1,473 @@
+        char dat[ElfPhdr::Size];
+        filein.read(dat, ElfPhdr::Size);
+        ElfPhdr phdr = ElfPhdr(dat);
+        elf_phdrs.push_back(phdr);
+        fileout.write(dat, ElfPhdr::Size);
+        at += ElfPhdr::Size;
+
+        filein.seekg(self_hdr.segment_info_offset + i * SegmentInfo::Size);
+        char segmentinfobuffer[SegmentInfo::Size];
+        filein.read(segmentinfobuffer, SegmentInfo::Size);
+        SegmentInfo segment_info = SegmentInfo(segmentinfobuffer);
+        segment_infos.push_back(segment_info);
+
+        if (segment_info.plaintext == SecureBool::NO)
+            encrypted = true;
+    }
+
+    std::vector<SceSegment> scesegs;
+
+    if (encrypted) {
+        scesegs = get_segments(filein, sce, SCE_KEYS, appinfo_hdr.sys_version, appinfo_hdr.self_type, npdrmtype, klictxt);
+    }
+
+    for (int i = 0; i < elf_hdr.e_phnum; i++) {
+        int idx;
+
+        if (!scesegs.empty())
+            idx = scesegs[i].idx;
+        else
+            idx = i;
+        if (elf_phdrs[idx].p_filesz == 0)
+            continue;
+
+        int pad_len = elf_phdrs[idx].p_offset - at;
+        if (pad_len < 0)
+            LOG_ERROR("ELF p_offset Invalid");
+
+        std::string padding;
+        for (int i = 0; i < pad_len; i++) {
+            padding += "00";
+        }
+
+        auto padding_array = string_utils::string_to_byte_array(padding);
+
+        std::string result(padding_array.begin(), padding_array.end());
+        fileout.write(result.c_str(), pad_len);
+
+        at += pad_len;
+
+        filein.seekg(segment_infos[idx].offset);
+        unsigned char *dat = new unsigned char[segment_infos[idx].size];
+        filein.read((char *)dat, segment_infos[idx].size);
+
+        unsigned char *decrypted_data = new unsigned char[segment_infos[idx].size];
+        if (segment_infos[idx].plaintext == SecureBool::NO) {
+            aes_context aes_ctx;
+            aes_setkey_enc(&aes_ctx, (unsigned char *)scesegs[i].key.c_str(), 128);
+            size_t ctr_nc_off = 0;
+            unsigned char ctr_stream_block[0x10];
+            aes_crypt_ctr(&aes_ctx, segment_infos[idx].size, &ctr_nc_off, (unsigned char *)scesegs[i].iv.c_str(), ctr_stream_block, dat, decrypted_data);
+        }
+
+        if (segment_infos[idx].compressed == SecureBool::YES) {
+            mz_stream stream;
+            stream.zalloc = Z_NULL;
+            stream.zfree = Z_NULL;
+            stream.opaque = Z_NULL;
+            stream.avail_in = 0;
+            stream.next_in = Z_NULL;
+            if (mz_inflateInit(&stream) != MZ_OK) {
+                LOG_ERROR("inflateInit failed while decompressing");
+                return;
+            }
+
+            std::string compressed_data((char *)decrypted_data, segment_infos[idx].size);
+            stream.next_in = (Bytef *)compressed_data.data();
+            stream.avail_in = compressed_data.size();
+
+            int ret;
+            char outbuffer[524288];
```

Comment:
Too large to stay on stack, you just got lucky. Make this on heap by using vector<char>;